extensions/interactions/CodeRepl/code-repl-prediction.service.ts

```diff
@@ -21,301 +21,324 @@
-
-        var tokenizedProgram = predictionService.getTokenizedProgramForCV(
-          pythonProgramTokens);
-        var programVector = CountVectorizerService.vectorize(
-          tokenizedProgram, cvVocabulary);
-
-        var predictionResult = SVMPredictionService.predict(
-          svmData, programVector);
-        if (predictionResult.predictionConfidence >
-            CODE_REPL_PREDICTION_SERVICE_THRESHOLD) {
-          return predictionResult.predictionLabel;
+      }
+    }
+
+    return tokenizedProgram;
+  }
+
+  getTokenizedProgramForCV(programTokens: any[]): any[] {
+    // Tokenize Python programs in dataset for winnowing.
+    const tokenizedProgram = [];
+
+    for (let i = 0; i < programTokens.length; i++) {
+      const token = programTokens[i];
+      const tokenId = token[0];
+      const tokenName = token[1];
+
+      if (
+        tokenId === this.PythonProgramTokenType.NL ||
+        tokenId === this.PythonProgramTokenType.COMMENT ||
+        tokenName.trim() === '') {
+        continue;
+      } else if (
+        tokenId === this.PythonProgramTokenType.NAME &&
+        this.KW_LIST.indexOf(tokenName) === -1) {
+        tokenizedProgram.push(this.TOKEN_NAME_VAR);
+      } else {
+        tokenizedProgram.push(tokenName);
+      }
+    }
+
+    return tokenizedProgram;
+  }
+
+  calcJaccardIndex(
+      multisetA: Array<number>, multisetB: Array<number>): number {
+    // Calculate jaccard index between two multisets.
+    multisetA.sort((x, y) => {
+      return x > y ? 1 : -1;
+    });
+    multisetB.sort((x, y) => {
+      return x > y ? 1 : -1;
+    });
+
+    const smallSet = (
+      (multisetA.length < multisetB.length) ?
+        multisetA.slice() : multisetB.slice());
+    let unionSet = (
+      (multisetA.length < multisetB.length) ?
+        multisetB.slice() : multisetA.slice());
+    let index = 0;
+    const extraElements = [];
+
+    smallSet.forEach(elem => {
+      while (index < unionSet.length && elem > unionSet[index]) {
+        index += 1;
+      }
+      if (index >= unionSet.length || elem < unionSet[index]) {
+        extraElements.push(elem);
+      } else if (elem === unionSet[index]) {
+        index += 1;
+      }
+    });
+
+    unionSet = unionSet.concat(extraElements);
+    if (unionSet.length === 0) {
+      return 0;
+    }
+
+    index = 0;
+    const intersectionSet = [];
+    multisetA.forEach(elem => {
+      while (index < multisetB.length && elem > multisetB[index]) {
+        index += 1;
+      }
+      if (index < multisetB.length && elem === multisetB[index]) {
+        intersectionSet.push(elem);
+        index += 1;
+      }
+    });
+
+    const coeff = intersectionSet.length / unionSet.length;
+    return coeff;
+  }
+
+  getProgramSimilarity(fingerprintA: any[], fingerprintB: any[]): number {
+    // Calculate similarity between two programs' fingerprints.
+    const multisetA = [];
+    const multisetB = [];
+
+    fingerprintA.forEach(hash => {
```

Comment:
Ditto and for every other occurrence like this below

Reply:
Done